atom/browser/browser_linux.cc

```diff
@@ -4,16 +4,67 @@
 
 #include "atom/browser/browser.h"
 
+#include <fcntl.h>
 #include <stdlib.h>
 
 #include "atom/browser/native_window.h"
 #include "atom/browser/window_list.h"
 #include "atom/common/atom_version.h"
+#include "base/command_line.h"
+#include "base/environment.h"
+#include "base/process/launch.h"
 #include "brightray/common/application_info.h"
+
+#if defined(USE_X11)
+#include "chrome/browser/ui/libgtkui/gtk_util.h"
 #include "chrome/browser/ui/libgtkui/unity_service.h"
+#endif
 
 namespace atom {
 
+const char kXdgSettings[] = "xdg-settings";
+const char kXdgSettingsDefaultBrowser[] = "default-web-browser";
+const char kXdgSettingsDefaultSchemeHandler[] = "default-url-scheme-handler";
+
+bool LaunchXdgUtility(const std::vector<std::string>& argv, int* exit_code) {
+  *exit_code = EXIT_FAILURE;
+  int devnull = open("/dev/null", O_RDONLY);
+  if (devnull < 0) return false;
+
+  base::LaunchOptions options;
+
+  base::FileHandleMappingVector remap;
+  remap.push_back(std::make_pair(devnull, STDIN_FILENO));
+  options.fds_to_remap = &remap;
+
+  base::Process process = base::LaunchProcess(argv, options);
+  close(devnull);
+
+  if (!process.IsValid())return false;
+  return process.WaitForExit(exit_code);
+}
+
+bool SetDefaultWebClient(const std::string& protocol) {
+#if defined(OS_CHROMEOS)
+  return true;
+#else
+  std::unique_ptr<base::Environment> env(base::Environment::Create());
+
+  std::vector<std::string> argv;
+  argv.push_back(kXdgSettings);
+  argv.push_back("set");
+  if (!protocol.empty()) {
```

Comment:
If the protocol is empty do we still want to continue?  If they didn't provide a protocol should we just stop trying?

Reply:
yeah i suppose we can immediately return false if there's no protocol, otherwise continue